client-sniffer/src/main/java/org/elasticsearch/client/sniff/Sniffer.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.client.sniff;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.http.HttpHost;
+import org.elasticsearch.client.RestClient;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Class responsible for sniffing nodes from an elasticsearch cluster and setting them to a provided instance of {@link RestClient}.
+ * Must be created via {@link Builder}, which allows to set all of the different options or rely on defaults.
+ * A background task fetches the nodes from elasticsearch and updates them periodically.
+ * Supports sniffing on failure, meaning that the client will notify the sniffer at each host failure, so that nodes can be updated
+ * straightaway.
+ */
+public final class Sniffer extends RestClient.FailureListener implements Closeable {
+
+    private static final Log logger = LogFactory.getLog(Sniffer.class);
+
+    private final boolean sniffOnFailure;
+    private final Task task;
+
+    private Sniffer(RestClient restClient, HostsSniffer hostsSniffer, long sniffInterval,
+                    boolean sniffOnFailure, long sniffAfterFailureDelay) {
+        this.task = new Task(hostsSniffer, restClient, sniffInterval, sniffAfterFailureDelay);
+        this.sniffOnFailure = sniffOnFailure;
+        restClient.setFailureListener(this);
+    }
+
+    @Override
+    public void onFailure(HttpHost host) throws IOException {
```

Comment:
I think this should be a 3rd class that can even be implemented by a user. It would also allow us to make the FailureListener a ctor argument. All we need to to is to add a `public void snif(HttpHost exclude)`  that the user can call. That way the can just have a `FailureListener` they can register on client creation and that is called on failure and therefor delegates to the sniffer? WDYT